Improve storage report

MendelMonteiro opened the pull request:
Add more statistics per message type to analyze which messages are being persisted the most.
Also simplify the `IReporter` API to use the same type for the input as we do for the output.

Open question, why do we even have the `TakeAndResetStorageReports` and `TakeAndResetReplaySpeedReports` methods as they are not used within Zebus at all?

ocoanet commented:
It is a good idea to add per message statistics. I agree that `TakeAndReset` methods should be part of `IReporter`, you can remove them.

Nitpicks:
- The name `IReporter` is way too generic in my opinion. It might be a good idea to rename it to `IPersistenceReporter`.
- I would rename `MessageTypeStatistics` to `MessageTypeStorageReport` and move it to its own file.
- `sendDurationInSeconds` and `ackDurationInSeconds` should be TimeSpan (and renamed accordingly).
- I would really add a computed property for `MessageBytes?.Length ?? 0`. 

On a side note, I noticed that the `SemaphoreSlim` is not disposed in `CqlStorage.Write`. It is a good thing, but it requires a comment. I will add it.

MendelMonteiro commented:
I'm not a huge fan of putting a one-line record in its own file especially as it is closely related to the `StorageReport` type.

I've deleted the `TakeAndReset` methods and implemented your other suggestions as well.

ltrzesniewski commented:
> I'm not a huge fan of putting a one-line record in its own file

Stop following the dogma of one type per file 🙂
